pkg/cloud/vsphere/context/cluster_context.go

```diff
@@ -0,0 +1,368 @@
+type ClusterContextParams struct {
+	Context    context.Context
+	Cluster    *clusterv1.Cluster
+	Client     client.ClusterV1alpha1Interface
+	CoreClient corev1.CoreV1Interface
+	Logger     logr.Logger
+}
+
+// ClusterContext is a Go context used with a CAPI cluster.
+type ClusterContext struct {
+	context.Context
+	Cluster       *clusterv1.Cluster
+	ClusterCopy   *clusterv1.Cluster
+	ClusterClient client.ClusterInterface
+	ClusterConfig *v1alpha1.VsphereClusterProviderConfig
+	ClusterStatus *v1alpha1.VsphereClusterProviderStatus
+	Logger        logr.Logger
+	client        client.ClusterV1alpha1Interface
+	machineClient client.MachineInterface
+	user          string
+	pass          string
+}
+
+// NewClusterContext returns a new ClusterContext.
+func NewClusterContext(params *ClusterContextParams) (*ClusterContext, error) {
+
+	parentContext := params.Context
+	if parentContext == nil {
+		parentContext = context.Background()
+	}
+
+	var clusterClient client.ClusterInterface
+	var machineClient client.MachineInterface
+	if params.Client != nil {
+		clusterClient = params.Client.Clusters(params.Cluster.Namespace)
+		machineClient = params.Client.Machines(params.Cluster.Namespace)
+	}
+
+	clusterConfig, err := v1alpha1.ClusterConfigFromCluster(params.Cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load cluster provider config")
+	}
+
+	clusterStatus, err := v1alpha1.ClusterStatusFromCluster(params.Cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load cluster provider status")
+	}
+
+	logr := params.Logger
+	if logr == nil {
+		logr = klogr.New().WithName("default-logger")
+	}
+	logr = logr.WithName(params.Cluster.APIVersion).WithName(params.Cluster.Namespace).WithName(params.Cluster.Name)
+
+	user := clusterConfig.VsphereUser
+	pass := clusterConfig.VspherePassword
+	if secretName := clusterConfig.VsphereCredentialSecret; secretName != "" {
+		if params.CoreClient == nil {
+			return nil, errors.Errorf("credential secret %q specified without core client", secretName)
+		}
+		logr.V(4).Info("fetching vsphere credentials", "secret-name", secretName)
+		secret, err := params.CoreClient.Secrets(params.Cluster.Namespace).Get(secretName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "error reading secret %q for cluster %s/%s", secretName, params.Cluster.Namespace, params.Cluster.Name)
+		}
+		user2, userOk := secret.Data[constants.VsphereUserKey]
+		pass2, passOk := secret.Data[constants.VspherePasswordKey]
+		if !userOk || !passOk {
+			return nil, errors.Wrapf(err, "improperly formatted secret %q for cluster %s/%s", secretName, params.Cluster.Namespace, params.Cluster.Name)
+		}
+		user, pass = string(user2), string(pass2)
+		logr.V(2).Info("found vSphere credentials")
+	}
+
+	return &ClusterContext{
+		Context:       parentContext,
+		Cluster:       params.Cluster,
+		ClusterCopy:   params.Cluster.DeepCopy(),
```

Comment:
Passing along a deep copy of a resource in a context seems odd, is this what CAPA does?

Reply:
Ah, I'm recalling our previous conversations now. Guessing it's needed to make the cluster patch request work?

Reply:
Sí. You compare to the original copy to generate the JSON patch.